### PR TITLE
Improve the type-utils package, split code, add some comments and examples

### PR DESCRIPTION
**Pull request by peter-sanderson**

### CODE REVIEW

File: packages/blockchain-link/src/workers/electrum/index.ts [hunk 1]
```diff
@@ -13,9 +14,6 @@ import type { ElectrumClient } from './client/electrum';
 type BlockListener = ReturnType<typeof L.blockListener>;
 type TxListener = ReturnType<typeof L.txListener>;
 
-// reason:
-// https://stackoverflow.com/questions/57103834/typescript-omit-a-property-from-all-interfaces-in-a-union-but-keep-the-union-s#answer-57103940
-type Without<T, K extends keyof any> = T extends any ? Omit<T, K> : never;
```
Comment (peter-sanderson) [inline packages/blockchain-link/src/workers/electrum/index.ts:18]: It is exactly same as [here](https://github.com/trezor/trezor-suite/pull/15743/files#diff-83c8cdefbce86aa426768d944fe5ec4b6b22f0d53bdbcef805e85a6233dab5e5R46).

File: packages/utils/src/mergeDeepObject.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 /* eslint-disable @typescript-eslint/no-use-before-define */
 // code shamelessly stolen from https://github.com/voodoocreation/ts-deepmerge
 
-type TAllKeys<T> = T extends any ? keyof T : never;
```
Comment (peter-sanderson) [inline packages/utils/src/mergeDeepObject.ts:4]: Same as [here](https://github.com/trezor/trezor-suite/pull/15743/files#diff-83c8cdefbce86aa426768d944fe5ec4b6b22f0d53bdbcef805e85a6233dab5e5R32)

File: packages/type-utils/src/utils.ts [hunk 3]
```diff
@@ -30,757 +65,46 @@ export type ConstWithOptionalFields<
     };
 };
 
-// Extract item from union
-export type ItemExtractor<M> = Extract<M, { type: M }>;
-
-// Unwrap type from Promise
-export type Await<T> = T extends {
```
Comment (peter-sanderson) [inline packages/type-utils/src/utils.ts:37]: unused

File: packages/type-utils/src/utils.ts [hunk 4]
```diff
@@ -30,757 +65,46 @@ export type ConstWithOptionalFields<
     };
 };
 
-// Extract item from union
-export type ItemExtractor<M> = Extract<M, { type: M }>;
```
Comment (peter-sanderson) [inline packages/type-utils/src/utils.ts:34]: unused

**Review by Lemonexe — APPROVED** [review]
:clap:

**Review by komret — APPROVED** [review]

**Pull request merged**